sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py

```diff
@@ -478,3 +487,189 @@ def test_splitting_missing_values(X_binned, all_gradients,
+        ([0] * 9 + [1] * 8, False, 2),
+
+        # only one category appear more than CAT_SMOOTH
+        ([0] * 12 + [1] * 8, False, 2),
+
+        # missing values + category appear less than CAT_SMOOTH
+        # 9 is missing
+        ([0] * 9 + [1] * 8 + [9] * 4, True, 2),
+
+        # 9 is missing
+        ([9] * 11, True, 0),
+    ])
+def test_splitting_categorical_no_splits(X_binned, has_missing_values,
+                                         n_bins_non_missing):
+    # Checks categorical splits are correct when there are no spliits
+
+    n_bins = max(X_binned) + 1
+    n_samples = len(X_binned)
+    X_binned = np.array([X_binned], dtype=X_BINNED_DTYPE).T
+    X_binned = np.asfortranarray(X_binned)
+
+    l2_regularization = 0.0
+    min_hessian_to_split = 1e-3
+    min_samples_leaf = 1
+    min_gain_to_split = 0.0
+
+    sample_indices = np.arange(n_samples, dtype=np.uint32)
+    all_gradients = np.ones(n_samples, dtype=G_H_DTYPE)
+    has_missing_values = np.array([False], dtype=np.uint8)
+    all_hessians = np.ones(1, dtype=G_H_DTYPE)
+    sum_gradients = all_gradients.sum()
+    sum_hessians = n_samples
+    hessians_are_constant = True
+
+    builder = HistogramBuilder(X_binned, n_bins, all_gradients,
+                               all_hessians, hessians_are_constant)
+
+    n_bins_non_missing = np.array([n_bins_non_missing], dtype=np.uint32)
+    monotonic_cst = np.array([MonotonicConstraint.NO_CST] * X_binned.shape[1],
+                             dtype=np.int8)
+    categorical = np.ones_like(monotonic_cst, dtype=np.uint8)
+    missing_values_bin_idx = n_bins - 1
+
+    splitter = Splitter(X_binned, n_bins_non_missing,
+                        missing_values_bin_idx, has_missing_values,
+                        monotonic_cst, categorical,
+                        l2_regularization, min_hessian_to_split,
+                        min_samples_leaf, min_gain_to_split,
+                        hessians_are_constant)
+
+    histograms = builder.compute_histograms_brute(sample_indices)
+    value = compute_node_value(sum_gradients, sum_hessians,
+                               -np.inf, np.inf, l2_regularization)
+    split_info = splitter.find_node_split(n_samples, histograms,
+                                          sum_gradients, sum_hessians, value)
+
+    # no split found
+    assert split_info.gain == -1
+
+
+def _assert_threshold_in_bitset(thresholds, bitset):
+    # bitset is assumed to be an array 4 of uint64
+    for threshold in thresholds:
+        i1 = threshold // 32
+        i2 = threshold % 32
+        assert (bitset[i1] >> np.uint32(i2)) & np.uint32(1)
+
+
+@pytest.mark.parametrize(
+    "X_binned, all_gradients, thresholds, n_bins_non_missing,"
+    "missing_values_bin_idx, has_missing_values",
+    [
+        # 3 categories (finds threshold by going left first)
+        ([0, 1, 2] * 11,  # X_binned
+         [1, 10, 1] * 11,  # all_gradients
+         [0, 2],  # thresholds
```

Comment:
Why are the missing values going to the right child? It seems like the left child is the one with the most samples?

Reply:
This was a mistake, now that the assert tests for equality, this is properly updated and tested.